sdk/src/main/java/com/bugsnag/android/Configuration.java

```diff
@@ -634,6 +645,33 @@ public void setDelivery(@NonNull Delivery delivery) {
         return map;
     }
 
+    /**
+     * Add a callback to modify or cancel a report immediately before
+     * it is delivered to Bugsnag.
+     *
+     * Unlike {@link BeforeNotify} callbacks, "before send" callbacks are not
+     * necessarily run in the same app session as when the report was generated,
+     * as the application may have terminated before delivery or networking
+     * conditions prevented delivering the report until a later date.
+     * <p>
+     * Example usage:
+     * <pre>
+     * config.beforeSend((Report report) -&gt; {
+     *   Error error = report.getError();
+     *   error.setContext(getImportantField(error));
+     *   return true;
+     * });
+     * </pre>
+     * @param beforeSend a callback to run before delivering errors to Bugsnag
+     * @see BeforeSend
+     * @see Report
+     */
+    public void beforeSend(BeforeSend beforeSend) {
+        if (!beforeSendTasks.contains(beforeSend)) {
```

Comment:
Good call on testing that order stays the same. I don't see why we would test that tasks are unique though. Why would it matter is someone adds the same callback multiple times?

Reply:
Due to lifecycle events the same callback could be unintentionally added multiple times, leading to confusion. Example: #127

Reply:
At that point people should fix their configuration to resolve it. For example, adding uniqueness checks would not prevent an identical callback from running multiple times if its added like:

```java
Bugsnag.beforeNotify(new BeforeNotify() {
    @Override
    public boolean run(Error error) { ... }
});
```

in a method which is called multiple times. There is no duplicate references involved to verify uniqueness.

Reply:
If I put the same beforeNotify function in multiple times I would expect it to run multiple times, so I'm confused why we would prevent that

Reply:
The activity lifecycle on Android makes it easy to add the same callback multiple times, accidentally. Consider the following code, which would add the same callback every time the user altered the device's orientation:

```
class MyActivity extends Activity implements BeforeNotify {

    // this can be called by the framework multiple times for the same instance
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState); object
        bugsnag.beforeNotify(this); // the activity object implements the callback
    }
}
```

bugsnag-android currently detects this scenario and prevents the same callback being added, by checking whether the collection already contains the object. It's worth noting that if the example code was altered to use an inner class instead, bugsnag-android would not prevent the addition:
```
bugsnag.beforeNotify(new BeforeNotify() {
    ...
});
```

This is because a new object would be created each time, and the previously added callback will not have referential equality with the new callback.

What sort of scenario would it be useful for the same callback to be invoked multiple times?